.clang-format

```diff
@@ -25,7 +25,7 @@ Standard: Cpp11
 PointerAlignment: Middle
 MaxEmptyLinesToKeep: 2
 KeepEmptyLinesAtTheStartOfBlocks: false
-AllowShortFunctionsOnASingleLine: InlineOnly
+AllowShortFunctionsOnASingleLine: All
```

Comment:
Will it reformat all short functions to single line?

Reply:
Yes, I thought that's the current format.

Reply:
clang-format is too unambiguous: if you specify that a function may reside in one line, it will convert all functions to one-liners if it fits within the line. The same for the line breaks inside a list of function arguments...

But according to our style, we allow short functions to be in one line, but don't make a preference of this style. It means that this change in clang-format will make things worse.

PS. Also we can note that `libs/consistent-hashing` was extracted from a separate codebase inside Yandex that does not necessarily conform to our style.

Reply:
> if you specify that a function may reside in one line, it will convert all functions to one-liners if it fits within the line.

Hmm, it doesn't for llvm master, it only merge lines when there are more than one statement.

btw, I've patched it https://github.com/amosbird/llvm-project/commit/2614cfcfb7e420d9396d509d1ca7934d165bbc01#diff-95ab41b4178f7f22290f22a09634eaadR1504

so that lambda works in our style. And it also works fine for oneliners (single statement)

Reply:
Probably I'm using outdated version of clang-format.

PS. The question: does this PR lowers the size of diff after applying clang-format to dbms directory?

Reply:
```
 original
-------------------------------------------------------------------------------
 Language            Files        Lines         Code     Comments       Blanks
-------------------------------------------------------------------------------
 C Header             1114       137099        95738        13681        27680
 C++                  1274       321708       270595         9770        41343
-------------------------------------------------------------------------------


 clang-format before
-------------------------------------------------------------------------------
 Language            Files        Lines         Code     Comments       Blanks
-------------------------------------------------------------------------------
 C Header             1114       136227        95957        13681        26589
 C++                  1274       321180       271564         9770        39846
-------------------------------------------------------------------------------

 clang-format after
-------------------------------------------------------------------------------
 Language            Files        Lines         Code     Comments       Blanks
-------------------------------------------------------------------------------
 C Header             1114       135743        95473        13681        26589
 C++                  1274       318299       268683         9770        39846
-------------------------------------------------------------------------------
```

btw, it catches a bug of my clang-format patch :) and it's now fixed

Reply:
I meant - the size of the diff that will be generated by applying clang-format. This measures how clang-format rules correspond to our existing code base.

Reply:
```
before
 2298 files changed, 53083 insertions(+), 54325 deletions(-)

after
 2309 files changed, 53155 insertions(+), 57764 deletions(-)
```

diff size is slightly larger than before

Reply:
It means that the new config is slightly less consistent with de-facto code style.

Reply:
Yup, however I would suggest that methods like this https://github.com/yandex/ClickHouse/blob/master/dbms/src/Access/AllowedClientHosts.cpp#L154 could be merged into one line.

Reply:
P.S, we can rely on this feature https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code to progressively migrate to clang-format

Reply:
> Yup, however I would suggest that methods like this https://github.com/yandex/ClickHouse/blob/master/dbms/src/Access/AllowedClientHosts.cpp#L154 could be merged into one line.

> And especially the empty ones https://github.com/yandex/ClickHouse/blob/master/dbms/src/Access/AllowedClientHosts.cpp#L138

It is a matter of taste. I found them Ok.

Reply:
> to progressively migrate to clang-format

Yes, but now we have to apply too many annotations to the code to make clang-format to work reasonable.

Reply:
> It is a matter of taste. I found them Ok.

 Ok. Somehow the style check doesn't care about this too  https://github.com/yandex/ClickHouse/blob/master/dbms/src/Access/AllowedClientHosts.cpp#L138 versus 
 https://github.com/yandex/ClickHouse/blob/master/dbms/src/Common/Macros.cpp#L15

> Yes, but now we have to apply too many annotations to the code to make clang-format to work reasonable.

After browsing through the ~200000 lines of diff code, I would say there are only a small fraction of code which need annotations. They are:

1. tabular like code in settings, brace initializer list, short function/method/trait/instantiation groups, switch case
2. some invocations of function like macros such as LOG
3. ternary operator alignment
4. long template invocation (e.g. std::is_same_v<loooooooooooooooooog condition>)

Current .clang-format also has other issues

Missing:

1. AfterCaseLabel: true
2. AllowShortLambdasOnASingleLine: All
3. SpaceInEmptyBlock: false

Inconsistent:

1. IndentPPDirectives: AfterHash or BeforeHash
2. AlignEscapedNewlines: Left or Right or False
3. AlignTrailingComments: True or False
4. AlignAfterOpenBracket: AlwaysBreak or Align